crml/eth-state-oracle/src/lib.rs

```diff
@@ -365,8 +374,36 @@ impl<T: Config> Module<T> {
 			request.caller,
 			caller_ss58_address
 		);
+
+		// state oracle precompile address
+		let state_oracle_precompile = T::StateOraclePrecompileAddress::get();
+		let state_oracle_precompile_ss58_address = T::AddressMapping::into_account_id(state_oracle_precompile);
+		// calculate required gas
+		let max_fee_per_gas = U256::from(T::MinGasPrice::get());
+		let max_priority_fee_per_gas = U256::one();
+
+		// `min_fee_per_gas` and `max_priority_fee_per_gas` are expressed in wei, scale to 4dp to work with CPAY amounts
+		let total_fee: Balance = scale_wei_to_4dp(
+			(max_fee_per_gas * request.callback_gas_limit + max_priority_fee_per_gas).saturated_into(),
+		);
+
+		// Exchange for fee asset if fee preferences have been set
+		if let Some(fee_preferences) = &request.fee_preferences {
+			let exchange_total = request.bounty + total_fee;
+			let max_payment = exchange_total.saturating_add(fee_preferences.slippage * exchange_total);
+			let exchange = FeeExchange::new_v1(fee_preferences.asset_id, max_payment);
+			T::BuyFeeAsset::buy_fee_asset(&caller_ss58_address, exchange_total, &exchange).map_err(|_| {
+				// Using general error to cover all cases due to fixed return type of pallet_evm::Error
+				Error::<T>::FeeExchangeFailed
+			})?;
```

Comment:
doesn't apply here, passing the error from buy_fee_asset should be fine
`T::BuyFeeAsset::buy_fee_asset(&caller_ss58_address, exchange_total, &exchange)?;`